src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/aarch64/macos/StackVaList.java

```diff
@@ -0,0 +1,282 @@
+
+    private static final VaList EMPTY = new SharedUtils.EmptyVaList(MemoryAddress.NULL);
+
+    private MemorySegment segment;
+    private final List<MemorySegment> attachedSegments;
+    private final MemorySegment livenessCheck;
+
+    private StackVaList(MemorySegment segment, List<MemorySegment> attachedSegments, MemorySegment livenessCheck) {
+        this.segment = segment;
+        this.attachedSegments = attachedSegments;
+        this.livenessCheck = livenessCheck;
+    }
+
+    public static final VaList empty() {
+        return EMPTY;
+    }
+
+    @Override
+    public int vargAsInt(MemoryLayout layout) {
+        return (int) read(int.class, layout);
+    }
+
+    @Override
+    public long vargAsLong(MemoryLayout layout) {
+        return (long) read(long.class, layout);
+    }
+
+    @Override
+    public double vargAsDouble(MemoryLayout layout) {
+        return (double) read(double.class, layout);
+    }
+
+    @Override
+    public MemoryAddress vargAsAddress(MemoryLayout layout) {
+        return (MemoryAddress) read(MemoryAddress.class, layout);
+    }
+
+    @Override
+    public MemorySegment vargAsSegment(MemoryLayout layout) {
+        return (MemorySegment) read(MemorySegment.class, layout);
+    }
+
+    @Override
+    public MemorySegment vargAsSegment(MemoryLayout layout, NativeScope scope) {
+        Objects.requireNonNull(scope);
+        return (MemorySegment) read(MemorySegment.class, layout, SharedUtils.Allocator.ofScope(scope));
+    }
+
+    private Object read(Class<?> carrier, MemoryLayout layout) {
+        return read(carrier, layout, MemorySegment::allocateNative);
+    }
+
+    private Object read(Class<?> carrier, MemoryLayout layout, SharedUtils.Allocator allocator) {
+        Objects.requireNonNull(layout);
+        SharedUtils.checkCompatibleType(carrier, layout, MacOsAArch64Linker.ADDRESS_SIZE);
+        Object res;
+        if (carrier == MemorySegment.class) {
+            TypeClass typeClass = TypeClass.classifyLayout(layout);
+            res = switch (typeClass) {
+                case STRUCT_REFERENCE -> {
+                    MemoryAddress structAddr = (MemoryAddress) VH_address.get(segment);
+                    try (MemorySegment struct = handoffIfNeeded(structAddr.asSegmentRestricted(layout.byteSize()),
+                         segment.ownerThread())) {
+                        MemorySegment seg = allocator.allocate(layout.byteSize());
+                        seg.copyFrom(struct);
+                        segment = segment.asSlice(VA_SLOT_SIZE_BYTES);
+                        yield seg;
+                    }
+                }
+                case STRUCT_REGISTER, STRUCT_HFA -> {
+                    MemorySegment struct = allocator.allocate(layout);
+                    struct.copyFrom(segment.asSlice(0L, layout.byteSize()));
+                    segment = segment.asSlice(VA_SLOT_SIZE_BYTES);
```

Comment:
Since arguments are packed according to alignment, I guess the offset could be larger or smaller than 8 bytes as well?

Reply:
This is using `alignUp(arg.layout.byteSize(), VA_SLOT_SIZE_BYTES)` in the writing code, so I think it should be the same here?

Reply:
Yes this is a mistake (`skip()` has the same problem). I'll add an extra case to VaListTest to catch it.